examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp

```diff
@@ -234,6 +234,201 @@ Status EnergyEvseDelegate::StartDiagnostics()
+    const DataModel::DecodableList<Structs::ChargingTargetScheduleStruct::DecodableType> & chargingTargetSchedules)
+{
+    /* A) check that the targets are valid
+     *  1) each target must be within valid range (TargetTimeMinutesPastMidnight < 1440)
+     *  2) each target must be within valid range (TargetSoC percent 0 - 100)
+     *      If SOC feature not supported then this MUST be 100 or not present
+     *  3) each target must be within valid range (AddedEnergy >= 0)
+     * B) Day of Week is only allowed to be included once
+     */
+
+    uint8_t dayOfWeekBitmap = 0;
+
+    auto iter = chargingTargetSchedules.begin();
+    while (iter.Next())
+    {
+        auto & entry    = iter.GetValue();
+        uint8_t bitmask = entry.dayOfWeekForSequence.GetField(static_cast<TargetDayOfWeekBitmap>(0x7F));
+        ChipLogProgress(AppServer, "DayOfWeekForSequence = 0x%02x", bitmask);
+
+        if ((dayOfWeekBitmap & bitmask) != 0)
+        {
+            // A bit has already been set - Return ConstraintError
+            ChipLogError(AppServer, "DayOfWeekForSequence has a bit set which has already been set in another entry.");
+            return Status::ConstraintError;
+        }
+        dayOfWeekBitmap |= bitmask; // add this day Of week to the previously seen days
+
+        auto iterInner   = entry.chargingTargets.begin();
+        uint8_t innerIdx = 0;
+        while (iterInner.Next())
+        {
+            auto & targetStruct          = iterInner.GetValue();
+            uint16_t minutesPastMidnight = targetStruct.targetTimeMinutesPastMidnight;
+            ChipLogProgress(AppServer, "[%d] MinutesPastMidnight : %d", innerIdx,
+                            static_cast<short unsigned int>(minutesPastMidnight));
+
+            if (minutesPastMidnight > 1439)
+            {
+                ChipLogError(AppServer, "MinutesPastMidnight has invalid value (%d)", static_cast<int>(minutesPastMidnight));
+                return Status::ConstraintError;
+            }
+
+            uint8_t targetSoC;
+            int64_t addedEnergy;
+
+            /* Check that we have at least an AddedEnergy or TargetSoC value */
+            if (!(targetStruct.targetSoC.HasValue()) && !(targetStruct.addedEnergy.HasValue()))
+            {
+                ChipLogError(AppServer, "Must have one of AddedEnergy or TargetSoC");
+                return Status::ConstraintError;
+            }
+
+            if (targetStruct.targetSoC.HasValue())
+            {
+                targetSoC = targetStruct.targetSoC.Value();
+                ChipLogProgress(AppServer, "[%d] TargetSoC           : %d", innerIdx, static_cast<int>(targetSoC));
+
+                if (targetSoC > 100)
+                {
+                    ChipLogError(AppServer, "TargetSoC has invalid value (%d)", static_cast<int>(targetSoC));
+                    return Status::ConstraintError;
+                }
+                /* If SoC feature is not supported check that the value is 100% */
```

Comment:
That's not what the spec says.  The spec says:

1. If SOC is supported, TargetSoC must be present (and then I have no idea what the conformance is for AddedEnergy, given how things are written).
2. If SOC not supported, then exactly one of TargetSoC and AddedEnergy must be present.

Reply:
Agree on 1/changed.

Wrt 2, the intent of the spec is:
- if SOC is supported, the TargetSoC is mandatory, and the AddedEnergy is optional (and the spec should probably be clarified to [SOC],O.a+ to say that)
- if SOC is not supported, one or both of TargetSoc and AddedEnergy must be present.
- And the code has been reworked to reflect this logic.